File: 2018/day02/day02.ts
/*
 * Problem 2 of the Advent-of-Code 2018
 */

import * as fs from 'fs';
import * as readline from 'readline';
export {}

const readInterface = readline.createInterface({
    input: fs.createReadStream('input.txt'),
    output: null
});

var input: Array<any> = []

readInterface.on('line', (line: string) => {
    input.push(line)
})

readInterface.on('close', () => {
    console.log("Part A: " + part_a(input).toString());
    console.log("Part B: " + part_b(input).toString());
})

function part_a(input_array: Array<string>): number {
    var num_twins: number = 0;
    var num_triplets: number = 0;

    /* For each input string */
    for (var i = 0; i < input_array.length; i++) {
        var count: Object = {};
        var id: string = input_array[i];
        /* Build up a count of each letter */
        for (var j = 0; j < id.length; j++) {
            var letter: string = id[j];
            if (!(letter in count)) {
                count[letter] = 1;
            } else {
                count[letter] += 1;
            }
        }

        /* Check if there are twins or triplets */
        if (are_there_twins(count)) {
            num_twins += 1;
        }
        if (are_there_triplets(count)) {
            num_triplets += 1;
        }
    }

    return num_twins * num_triplets;
}

/* Returns true if the object has a [key, value] pair where the value == 2 */
function are_there_twins(dict: Object): boolean {
    for (const key in dict) {
        if (dict[key] == 2) {
            return true;
        }
    }
    return false;
}

/* Returns true if the object has a [key, value] pair where the value == 3 */
function are_there_triplets(dict: Object): boolean {
    for (const key in dict) {
        if (dict[key] == 3) {
            return true;
        }
    }
    return false;
}

function part_b(input_array: Array<string>): number {
    return 0;
}

File: 2021/day02/day02.py
"""
Problem 2 of the Advent-of-Code 2019
"""

from typing import Any, List

def read_inputs(filename: str) -> List[Any]:
    input_list = []
    with open(filename, 'r') as fp:
        for line in fp:
            input_list.append(line.strip())
    return input_list

def part_a(input_list: List[int]) -> int:
    z = 0
    x = 0
    for i in input_list:
        if 'forward' in i:
            x += int(i.split()[-1])
        if 'up' in i:
            z -= int(i.split()[-1])
        if 'down' in i:
            z += int(i.split()[-1])

    return x * z

def part_b(input_list: List[int]) -> int:
    aim = 0
    z = 0
    x = 0
    for i in input_list:
        if 'forward' in i:
            x += int(i.split()[-1])
            z += (int(i.split()[-1]) * aim)
        if 'up' in i:
            aim -= int(i.split()[-1])
        if 'down' in i:
            aim += int(i.split()[-1])

    return x * z

if __name__ == "__main__":
    input_list = read_inputs('input.txt')
    print(f'Part A: {part_a(input_list)}')
    print(f'Part B: {part_b(input_list)}')
